File: py/other/classNode.py
class Queue(object):
  def __init__(self):
    self.items = []
  def isEmpty(self):
    # len(self.items) == 0
    # print "inside the isEmpty method returns ", self.items == []
    return self.items == []
    # return len(self.items) == 0
  def enqueue(self, item):
    self.items.insert(0, item)
  def dequeue(self):
    if self.isEmpty() is not True:
      return self.items.pop()
  def size(self):
    return len(self.items)
  def printQueue(self):
    print self.items


class Node(object):
  def __init__(self, data, left = None, right = None):
    self.data = data
    self.left = left
    self.right = right
  def inOrder(self):
    # print "class method inOrder", self.data
    if self.left is not None:
      self.left.inOrder()
    print self.data
    if self.right is not None:
      self.right.inOrder()
  def preOrder(self):
    # print "class method preOrder", self.data
    print self.data
    if self.left is not None:
      self.left.preOrder()
    if self.right is not None:
      self.right.preOrder()
  def postOrder(self):
    # print "class method postOrder", self.data
    if self.left is not None:
      self.left.postOrder()
    if self.right is not None:
      self.right.postOrder()
    print self.data
  def levelOrder(self):
    q = Queue()
    q.enqueue(self)
    while q.isEmpty() is not True:
      a = q.dequeue()
      print a.data
      if a.left is not None:
        q.enqueue(a.left)
      if a.right is not None:
        q.enqueue(a.right)




# qq = Queue()
# qq.enqueue("1")
# qq.enqueue("2")
# qq.printQueue()
# print qq.dequeue()
# qq.printQueue()

myNode = Node("a", Node("b", Node("d"), Node("e")), Node("c", Node("f"), Node("g")))

# stand alone function
def inOrderPrint(node):
  if node is not None:
    # print "stand alone function inOrderPrint", node.data
    inOrderPrint(node.left)
    print node.data
    inOrderPrint(node.right)

def levelOrderPrint(node):
  if node is not None:
    q = Queue()
    q.enqueue(node)
    while q.isEmpty() is not True:
      node = q.dequeue()
      print node.data
      if node.left is not None:
        q.enqueue(node.left)
      if node.right is not None:
        q.enqueue(node.right)

levelOrderPrint(myNode)
print "in level order SA", "8"*20


myNode.inOrder()
print "in order ", "*"*20
myNode.preOrder()
print "pre order ", "*"*20
myNode.postOrder()
print "post order ", "*"*20
inOrderPrint(myNode)
print "in order SA", "8"*20
myNode.levelOrder()
print "in level order ", "*"*20

File: js/other/classNode.js
"use strict";
class Queue{
  constructor(queue = []){
    this.queue = queue;
  }
  isEmpty(){
    return this.queue.length === 0;
  }
  enqueue(item){
    this.queue.unshift(item);
  }
  dequeue(){
    if (!this.isEmpty()){
      return this.queue.pop();
    }
  }
  size(){
    return this.queue.length;
  }
}


class Node{
  constructor (data, left = null, right = null){
    this.data = data;
    this.left = left;
    this.right = right;
  }
  // depth first search
  inOrder() {
    // console.log("inOrder method");
    if (this.left){
      this.left.inOrder();
    }
    console.log(this.data);
    if (this.right){
      this.right.inOrder();
    }
  }
  preOrder() {
    // console.log("preOrder method");
    console.log(this.data);
    if (this.left) {
      this.left.preOrder();
    }
    if (this.right){
      this.right.preOrder();
    }
  }
  postOrder(){
    // console.log("postOrder method");
    if (this.left) {
      this.left.postOrder();
    }
    if (this.right){
      this.right.postOrder();
    }
    console.log(this.data);
  }
  // breadth first search
  levelOrder(){
    let q = new Queue();
    let node;
    q.enqueue(this);
    // console.log(q);
    // console.log(!q.isEmpty())
    while (!q.isEmpty()){
      node = q.dequeue();
      console.log(node.data);
      if (node.left){
        q.enqueue(node.left);
      }
      if (node.right){
        q.enqueue(node.right);
      }
    }
  }
}
/*
let q = new Queue();
console.log(q.size());
q.enqueue("1");
q.enqueue("2");
console.log(q);
q.dequeue();
console.log(q);
*/
let myNode = new Node("a", new Node("b"), new Node("c"));
// console.log(myNode);
console.log("in order");
myNode.inOrder();
console.log("pre order");
myNode.preOrder();
console.log("post order");
myNode.postOrder();
console.log("in level order");
myNode.levelOrder();